rasa/shared/core/domain.py

```diff
@@ -737,25 +762,47 @@ def add_knowledge_base_slots(self) -> None:
                     self.slots.append(TextSlot(s, influence_conversation=False))
 
     def index_for_action(self, action_name: Text) -> Optional[int]:
-        """Look up which action index corresponds to this action name."""
-
+        """Looks up which action index corresponds to this action."""
         try:
-            return self.action_names.index(action_name)
+            return self.action_names_or_texts.index(action_name)
         except ValueError:
             self.raise_action_not_found_exception(action_name)
 
-    def raise_action_not_found_exception(self, action_name) -> NoReturn:
-        action_names = "\n".join([f"\t - {a}" for a in self.action_names])
+    def raise_action_not_found_exception(self, action_name_or_text: Text) -> NoReturn:
+        """Raises exception if action name or text not part of the domain.
+
+        Args:
+            action_name_or_text: Name of an action or its text in case it's an
+                end-to-end bot utterance.
+
+        Raises:
+            ActionNotFoundException: If `action_name_or_text` are not part of this
+                domain.
+        """
+        action_names = "\n".join([f"\t - {a}" for a in self.action_names_or_texts])
         raise ActionNotFoundException(
-            f"Cannot access action '{action_name}', "
+            f"Cannot access action '{action_name_or_text}', "
             f"as that name is not a registered "
             f"action for this domain. "
             f"Available actions are: \n{action_names}"
         )
 
     def random_template_for(self, utter_action: Text) -> Optional[Dict[Text, Any]]:
+        """Returns a random response for an action name.
+
+        Args:
+            utter_action: The name of the utter action.
+
+        Returns:
+            A response for an utter action.
+        """
         import numpy as np
 
+        rasa.shared.utils.io.raise_warning(
```

Comment:
likewise with `raise_deprecation_warning`

Reply:
@wochinge could you please take this over

Reply:
See comment [here](https://github.com/RasaHQ/rasa/pull/7496#discussion_r542290872). Also added a comment in the code.